src/js/utils/ajax.js

```diff
@@ -152,7 +165,8 @@ function _readyStateChangeHandler(options) {
                 } else {
                     message = '' + xhr.status + '(' + xhr.statusText + ')';
                 }
-                return options.onerror(message, options.url, xhr);
+                _error(options, message, (xhr.status < 600) ? xhr.status : ERROR_XHR_UNKNOWN);
```

Comment:
This error handling seems odd, we only handle the case where the code is 200 or a value greater than 400. Is that because the state change handler gets called several times with a ready state of 4 until status is 200 or > 400? If not, I could rewrite in a separate PR as:
```
if (xhr.readyState === 4) {
            clearTimeout(options.timeoutId);
            if (xhr.status === 200) {
                return _ajaxComplete(options)(e);
            }
            let message;
            if (xhr.status >= 400) {
                if (xhr.status === 404) {
                    message = 'File not found';
                } else {
                    message = '' + xhr.status + '(' + xhr.statusText + ')';
                }
            }
            _error(options, message, (xhr.status < 600) ? xhr.status : ERROR_XHR_UNKNOWN);
}
```
that way we can catch a broader status range, i.e. xhr.status is undefined

Reply:
> This error handling seems odd, we only handle the case where the code is 200 or a value greater than 400. Is that because the state change handler gets called several times with a ready state of 4 until status is 200 or > 400?

`if (xhr.readyState === 4) {` handles waiting for the request to be complete.

When it is, code `200` signifies that the response was OK, and we can use it. Anything else that is not an error would be handled by XHR (redirect), or by triggering the onerror callback.

`400-599` are client-server errors, and we want to know when these happen for a file because it helps us tell why the client couldn't retrieve the file. Other status codes like I mentioned would have been handled already so not helpful.